### PR TITLE
[WIP] Alt implementation for #24343

### PR DESCRIPTION
**Pull request by jarshwah**
Alternative implementation for https://github.com/django/django/pull/4134

### CODE REVIEW

**Comment by akaariai** [comment]
I think this is the right approach. But instead of using the target, lets redefine outputfield for col. So, output field always tells the database returned value's type, but it isn't necessarily the field with correct column attribute. This way we can alwaya use the output_field argument consistently.

Could we just change the charfield foreign key test to return empty string on Oracle? If we ever get to split foreign key to relation and underlying concrete field the current behavior will be near impossible to support.

**Comment by jarshwah** [comment]
@akaariai I don't see how we can redefine output_field for Col. It's used for fetching the related_field information (for ForeignKeys), and then running all applicable converters. It's only when generating the column attname (field_id) that you need the target. Unless I've missed something?

I don't think it's possible to change oracle foreign key char fields in the backend converter because the converters don't know that they are working with a foreign key at that point. Hence @mjtamlyn's work-around to redefine all converters within the ForeignKey (including backend converters), and my workaround of applying the target field converter within Col expression.

But I don't like the brittleness of applying the target converter within Col. If there's another option I'm very open to it - but I couldn't come up with a nicer way just yet. Ideas?

**Comment by akaariai** [comment]
Ok, I got confused with output_field and target for Col. The way it is is very good (except perhaps we should change the source parameter name to output_field name now that we still can for col and get_col()?)

For the varchar foreign key case, my point is that we shouldn't need any special cased converter for ForeignKeys. Return whatever the related field's converter will return.

**Comment by jarshwah** [comment]
> For the varchar foreign key case, my point is that we shouldn't need any special cased converter for ForeignKeys. Return whatever the related field's converter will return.

For Oracle that returns the empty string for null foreign keys that point to string like fields. Null foreign keys should return None on all backends - it's what the tests expect. The backend converter (operating on the remote field type) see's a value of None, but since it's a charfield Oracle will convert it to "". When it finally reaches the user it's an empty string instead of None.

Agree with changing the name to output_field. Will do that tomorrow.

**Comment by shaib** [comment]
@akaariai:
> For the varchar foreign key case, my point is that we shouldn't need any special cased converter for ForeignKeys. Return whatever the related field's converter will return.

To expand on @jarshwah's response: The value of a FK attribute should be either None or an instance of the pointed-to model; an empty (or any) string would be a type error.

File: django/db/models/fields/related.py [hunk 1]
```diff
@@ -2051,6 +2052,17 @@ def db_type(self, connection):
...
         return {"type": self.db_type(connection), "check": []}
 
+    def from_db_value(self, value, connection, context):
+        if ((not value) and isinstance(value, six.string_types) and
+            connection.features.interprets_empty_strings_as_nulls):
+                return None
+        return value
+
+    def get_col(self, alias, source=None):
+        if source is None:
```
Comment (shaib) [inline django/db/models/fields/related.py]: Can `source` be falsey and not `None`? If not, I think it's better to use `source or self.related_field` and make the method a one-liner.

File: tests/model_fields/models.py [hunk 2]
```diff
@@ -369,6 +369,10 @@ class PrimaryKeyUUIDModel(models.Model):
     id = models.UUIDField(primary_key=True, default=uuid.uuid4)
 
 
+class RelatedToUUIDModel(models.Model):
+    rel = models.ForeignKey('PrimaryKeyUUIDModel')
```
Comment (timgraham) [inline tests/model_fields/models.py]: A name like `uuid_pk` might avoid confusion with `field.rel` but no worries if you think the current name is fine.

**Comment by jarshwah** [comment]
I'm not going to be available for the next few days. I'm happy with the patch at this point though - so if someone would like to merge and backport, that'd be nice. @akaariai are you OK with the FK converter based on what Shai and I had to say about it? There's probably more that can be done given enough time and thought, but maybe not for 1.8 beta.

**Comment by jarshwah** [comment]
Not terribly related to this patch, but maybe we can look into redesigning how converters work. We already have a basic pipeline (backend, field, expression), but we have no way of exiting the pipeline early, which I think would be extremely handy.

```
for converter in converters:
    if not continue: 
        break
    value, continue = converter(value)
```

This would help for ForeignKeys, because we could slip the foreign key converter into the beginning of the pipeline, and exit early if the database value was None. At the moment we don't have a mechanism for controlling the order or early exit though, but I think it could be a handy tool to have.

**Comment by akaariai** [comment]
Yes, something like that would be useful. If I'm not mistaken, now every backend will need a converter for every foreign key field, just because Oracle needs this for CharField foreign keys. And we can't do this as a backend converter, because CharField's converter will override what we did in the backend converter.

Actually I think we could do a bit less than what we do currently for ForeignKeys. Instead of providing from_db_value(), we could override ForeignKey's get_db_converters(), and add a converter to super() converters only if the backend needs one (possibly also only if the target field is CharField). No need to run the no-op converters on non-Oracle backends for every row's every foreign key column.

Another fix would be fetching all the converters in the expressions get_db_converters() method. We can't do this easily currently as backend converters have different signature than field converters. Maybe we should change this, providing a redundant arguments for converters is maybe less ugly than having different signatures?

File: django/db/models/fields/related.py [hunk 3]
```diff
@@ -2064,6 +2064,15 @@ def db_type(self, connection):
     def db_parameters(self, connection):
         return {"type": self.db_type(connection), "check": []}
 
+    def from_db_value(self, value, connection, context):
```
Comment (mjtamlyn) [inline django/db/models/fields/related.py]: This should be done in `get_db_converters` not as `from_db_value` for performance reasons. Where possible, we should avoid the existence of any converter functions, even when they become noops like this.

**Comment by mjtamlyn** [comment]
@jarshwah I'm not convinced allowing early exit from the chain would be advisable, bear in mind that `from_db_value` (a converter) is public API. Someone can make a subclass of any existing field, add `from_db_value` and expect it to be run.

@akaariai I think providing all the arguments now would make more sense given the change of arguments passed - it would have been really odd passing "field" to a converter on a field, but we are not passing "expression" to those converters, and "connection" to those on a backend's ops module, which is rather less pointless. I'll see if I can tidy all this up.

**Comment by akaariai** [comment]
I think we can merge this as is, and then later on add a get_db_converters() for ForeignKeys, and remove the from_db_value().

**Pull request closed without merging**